uniffi_bindgen/src/bindings/gecko_js/gen_gecko_js.rs

```diff
@@ -0,0 +1,463 @@
+            // to instantiate `ViaFfi`, not `nsString`.
+            Type::String => "nsAString".into(),
+            Type::Optional(inner) => match inner.as_ref() {
+                Type::String => "nsAString".into(),
+                _ => in_arg_type_cpp(type_)?,
+            },
+            _ => in_arg_type_cpp(type_)?,
+        };
+        let detail = detail_cpp(namespace)?;
+        Ok(format!(
+            "{}::ViaFfi<{}, {}>::Lower({})",
+            detail,
+            lifted,
+            type_ffi(&FFIType::from(type_))?,
+            from
+        ))
+    }
+
+    /// Generates an expression for lifting a C return type from the FFI into a
+    /// C++ out parameter.
+    pub fn lift_cpp(
+        namespace: &str,
+        type_: &Type,
+        from: &str,
+        into: &str,
+    ) -> Result<String, askama::Error> {
+        let lifted = match type_ {
+            // Out arguments are also `nsAString`, so we need to use it for the
+            // instantiation.
+            Type::String => "nsAString".into(),
+            Type::Optional(inner) => match inner.as_ref() {
+                Type::String => "nsAString".into(),
+                _ => type_cpp(type_)?,
+            },
+            _ => type_cpp(type_)?,
+        };
+        let detail = detail_cpp(namespace)?;
+        Ok(format!(
+            "{}::ViaFfi<{}, {}>::Lift({}, {})",
+            detail,
+            lifted,
+            type_ffi(&FFIType::from(type_))?,
+            from,
+            into,
+        ))
+    }
+
+    pub fn var_name_webidl(nm: &dyn fmt::Display) -> Result<String, askama::Error> {
+        Ok(nm.to_string().to_mixed_case())
+    }
+
+    pub fn enum_variant_webidl(nm: &dyn fmt::Display) -> Result<String, askama::Error> {
+        Ok(nm.to_string().to_mixed_case())
+    }
+
+    pub fn class_name_webidl(nm: &dyn fmt::Display) -> Result<String, askama::Error> {
+        Ok(nm.to_string().to_camel_case())
+    }
+
+    pub fn class_name_cpp(nm: &dyn fmt::Display) -> Result<String, askama::Error> {
+        Ok(nm.to_string().to_camel_case())
+    }
+
+    pub fn fn_name_webidl(nm: &dyn fmt::Display) -> Result<String, askama::Error> {
+        Ok(nm.to_string().to_mixed_case())
+    }
+
+    /// For interface implementations, function and methods names are
+    /// UpperCamelCase, even though they're mixedCamelCase in WebIDL.
+    pub fn fn_name_cpp(nm: &dyn fmt::Display) -> Result<String, askama::Error> {
+        Ok(nm.to_string().to_camel_case())
+    }
+
+    pub fn field_name_cpp(nm: &str) -> Result<String, askama::Error> {
+        Ok(format!("m{}", nm.to_camel_case()))
+    }
+
+    pub fn enum_variant_cpp(nm: &dyn fmt::Display) -> Result<String, askama::Error> {
+        // TODO: Make sure this does the right thing for hyphenated variants.
+        // Example: "bookmark-added" should become `Bookmark_added`, because
+        // that's what Firefox's `Codegen.py` spits out.
```

Comment:
I filed https://github.com/mozilla/uniffi-rs/issues/294 for this, but the consequences is that hyphenated names will produce a page-long compilation error because the enum variant that UniFFI generates doesn't exist in the variants that `Codegen.py` spits out 😬